core/state.py

```diff
@@ -1161,14 +1162,33 @@ def __enter__(self):
     def __exit__(self, type, value_, traceback):
         # type: (Any, Any, Any) -> None
         if self.vars is not None:
-            self.mem.var_stack.pop()
+            self._Pop()
 
         if self.pos_args is not None:
             self.mem.argv_stack.pop()
 
         if self.dollar0 is not None:
             self.mem.SetLocalName(self.dollar0_lval, value.Undef)
 
+    # Note: _Push and _Pop are separate methods because the C++ translation
+    # doesn't like when they are inline in __init__ and __exit__.
+    def _Push(self, pairs):
```

Comment:
How about creating free functions like:

```
state.PushVars(mem, restore)
state.PopVars(mem, restore)
```

and then using it from both `ctx_Shvar` and `ctx_Eval` ?

---

Also, I think where this is going is that we are going to move it up to the YSH level:


```
push-vars ({a: 42, b: 'foo'} {
  echo $a $b
}
```

There is a thread about "unifying stacks with context managers" on Zulip

Reply:
What about using `ctx_Eval` instead of `ctx_Shvar` for `shvar`?

Reply:
Oh yes, that makes sense!  I wasn't sure how similar they were